File: JavaScript/3.4 Test doubles/IncomeCalculator.js
const Position = {
    NONE: 'NONE',
    BOSS: 'BOSS',
    PROGRAMMER: 'PROGRAMMER',
    SURFER: 'SURFER'
};

class ICalcMethod {
    calc() { }
}

class IncomeCalculator {
    constructor() {
        this.calcMethod = undefined;
        this.position = undefined;
    }

    setCalcMethod(calcMethod) {
        this.calcMethod = calcMethod;
    }

    setPosition(position) {
        this.position = position;
    }

    calc() {
        if (this.calcMethod == undefined) {
            throw new Error("Calc method not yet maintained");
        }

        if (this.position == undefined) {
            throw new Error("Position not yet maintained");
        }

        return this.calcMethod.calc(this.position);
    }
}

module.exports = {
    Position,
    ICalcMethod,
    IncomeCalculator,
};

File: Solutions/C#/4.5 - Test doubles/QACalculator/IncomeCalculator.cs
using System;

namespace QACalculator
{
    public class IncomeCalculator
    {
        private ICalcMethod calcMethod;
        private Position position;

        public void SetCalcMethod(ICalcMethod calcMethod)
        {
            this.calcMethod = calcMethod;
        }

        public void SetPosition(Position position)
        {
            this.position = position;
        }

        public double Calc()
        {
            if (calcMethod == null)
            {
                throw new Exception("CalcMethod not yet maintained");
            }
            if (position == Position.NONE)
            {
                throw new Exception("Position not yet maintained");
            }
            return calcMethod.Calc(position);
        }
    }
}
